lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +175,99 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(
+    table,
+    columns,
+    name_suffix:,
+    if_not_exists: nil,
+    using: nil,
+    unique: nil,
+    where: nil,
+    comment: nil
+  )
+    raise ArgumentError, "Expected <name_suffix> to be present" unless name_suffix.present?
+
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    parent_schema, parent_table = _schema_and_table_for(table)
+
+    fully_qualified_parent_table = "#{connection.quote_schema_name(parent_schema)}.#{parent_table}"
+    parent_index = "#{parent_table}_#{name_suffix}"
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{parent_table.inspect} is not a partitioned table" unless _partitioned_table?(parent_schema, parent_table)
+
+    _validate_index_name!(parent_index)
+
+    child_schemas_and_tables = connection.select_rows(<<~SQL)
+      SELECT child_ns.nspname, child.relname
+      FROM pg_inherits
+        JOIN pg_class parent        ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child         ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace parent_ns ON parent.relnamespace = parent_ns.oid
+        JOIN pg_namespace child_ns  ON child.relnamespace = child_ns.oid
+      WHERE parent.relname = #{connection.quote(parent_table)}
+        AND parent_ns.nspname = #{connection.quote(parent_schema)}
+    SQL
+
+    child_tables_with_metadata = child_schemas_and_tables.map do |child_schema, child_table|
+      raise PgHaMigrations::InvalidMigrationError, "Partitioned table #{parent_table.inspect} contains sub-partitions" if _partitioned_table?(child_schema, child_table)
+
+      child_index = "#{child_table}_#{name_suffix}"
+
+      _validate_index_name!(child_index)
+
+      [child_schema, child_table, child_index]
+    end
+
+    # TODO: take out ShareLock after issue #39 is implemented
+    safely_acquire_lock_for_table(fully_qualified_parent_table) do
+      # CREATE INDEX ON ONLY parent_table
```

Comment:
I think the main goal of this method is to be able to concurrently create an index when data already exists in a partitioned table. I could maybe see adding a warning message that says, "hey this is gonna take awhile", if there's a ton of data. But we don't really have a precedent for that, as there are plenty of other operations in this library that could take a really long time

For this particular call within this method, I _don't think_ the size of the data really impacts the speed in which `CREATE INDEX ON ONLY` executes

Reply:
Oh wait, are we guaranteeing this is a natively partitioned table already (and not a partman partitioned one)? If so, I think we'd be fine, because the "partition" table _when natively partitioned_ is [guaranteed to be empty](https://www.postgresql.org/docs/current/ddl-partitioning.html):

> The partitioned table itself is a “virtual” table having no storage of its own.

But if we support partman partitioned tables than the partition parent itself can have data, and we're not running this `concurrently`, so it could be a long-held lock.

Reply:
Yeah, we're querying `pg_partitioned_table` so this method will only target natively partitioned tables. I actually don't know if attaching an index is something you can do with normal table inheritance... But either way, I think it's fine to just target natively partitioned tables in this iteration (the readme already calls this out)